pkg/pgmodel/ingestor/copier.go

```diff
@@ -124,22 +147,28 @@ func persistBatch(conn pgxconn.PgxConn, sw *seriesWriter, elf *ExemplarLabelForm
 		return fmt.Errorf("copier: formatting exemplar label values: %w", err)
 	}
 
-	doInsertOrFallback(conn, insertBatch...)
+	doInsertOrFallback(ctx, conn, insertBatch...)
 	return nil
 }
 
-func copierGetBatch(batch []readRequest, in <-chan readRequest) ([]readRequest, bool) {
+func copierGetBatch(ctx context.Context, batch []readRequest, in <-chan readRequest) ([]readRequest, bool) {
+	_, span := tracer.Default().Start(ctx, "get-batch")
+	defer span.End()
 	//This mutex is not for safety, but rather for better batching.
 	//It guarantees that only one copier is reading from the channel at one time
 	//This ensures bigger batches as well as less spread of a
 	//single http request to multiple copiers, decreasing latency via less sync
 	//overhead especially on low-pressure systems.
+	span.AddEvent("Acquiring lock")
 	getBatchMutex.Lock()
+	span.AddEvent("Got the lock")
 	defer getBatchMutex.Unlock()
+	defer span.AddEvent("Unlocking")
```

Comment:
Why not combine these two defers into one?

Also, Go question, are defers executed in reverse order? i.e. is the call to `span.End()` guaranteed to happen AFTER the call to `span.AddEvent(...)`?

Reply:
Right, I guess I can combine them, need to be careful with closures there. Defers are called last-to-first so End will always come after AddEvent in this case.